### ISSUE TITLE
tensor parallelism exception

### ISSUE THREAD

**Issue by quarterturn** [bigscience-workshop/petals#246]
Dual P100 setup:

```
(petals) qtr@pve:~/petals$ python -m petals.cli.run_server bigscience/bloomz-petals --tensor_parallel_devices cuda:0 cuda:1 --num_blocks 6
Jan 31 07:28:41.703 [INFO] Running Petals 1.1.2
Jan 31 07:28:56.893 [INFO] Direct reachability: 3/5
Jan 31 07:28:56.937 [INFO] This server will run DHT in full peer mode
Jan 31 07:28:59.927 [INFO] Connecting to the public swarm, peer_id = 12D3KooWMMiL1fw5oPpsegj4E1iu7Woqqc9rLzjrGoannTdJNxY1
Jan 31 07:29:00.045 [INFO] Model weights will be split between cuda:0, cuda:1
Jan 31 07:29:00.082 [WARN] [/home/qtr/anaconda3/envs/petals/lib/python3.10/site-packages/petals/server/server.py.__init__:167] Tensor parallelism doesn't work properly with 8-bit weights yet, loading weights in 16-bit. You can explicitly set `--load_in_8bit True` to override this
Jan 31 07:29:00.082 [INFO] Model weights will be loaded in auto format
Jan 31 07:29:00.082 [INFO] Attention cache for all blocks will consume up to 3.00 GiB
Jan 31 07:29:00.084 [INFO] Loading throughput info
Jan 31 07:29:01.221 [INFO] Announced that blocks [7, 8, 9, 10, 11, 12] are joining
Jan 31 07:29:01.622 [INFO] Reachability service started
Jan 31 07:29:29.358 [INFO] Loaded bigscience/bloomz-petals block 7, <All keys matched successfully>
Jan 31 07:30:04.819 [INFO] Loaded bigscience/bloomz-petals block 8, <All keys matched successfully>
Jan 31 07:30:35.309 [INFO] Loaded bigscience/bloomz-petals block 9, <All keys matched successfully>
Jan 31 07:31:05.673 [INFO] Loaded bigscience/bloomz-petals block 10, <All keys matched successfully>
Downloading: 100%|█████████████████████████████████████████████████████████████████████████████| 4.93G/4.93G [03:15<00:00, 25.2MB/s]
Jan 31 07:34:51.803 [INFO] Loaded bigscience/bloomz-petals block 11, <All keys matched successfully>
Jan 31 07:35:30.357 [INFO] Loaded bigscience/bloomz-petals block 12, <All keys matched successfully>
Jan 31 07:35:34.023 [INFO] Server is reachable from the Internet. It will appear at http://health.petals.ml soon
Jan 31 07:35:35.407 [INFO] Started
Jan 31 08:16:10.959 [INFO] rpc_inference.open(blocks=7:13, remote_peer=...AM4oxF)
Jan 31 08:16:10.962 [INFO] rpc_inference.wait_for_alloc(size=0.16 GiB), already used 0.00/3.00 GiB (0.0%)
Jan 31 08:16:10.965 [INFO] rpc_inference.alloc(size=0.16 GiB)
Jan 31 08:16:11.371 [ERROR] [hivemind.moe.server.runtime.run:104] Caught indices should be either on cpu or on the same device as the indexed tensor (cuda:1), attempting to recover
Traceback (most recent call last):
  File "/home/qtr/anaconda3/envs/petals/lib/python3.10/site-packages/hivemind/moe/server/runtime.py", line 91, in run
    outputs = pool.process_func(*batch)
  File "/home/qtr/anaconda3/envs/petals/lib/python3.10/site-packages/petals/server/backend.py", line 175, in __call__
    (hidden_states,) = self.backends[inference_info.uid].inference_step(hidden_states, hypo_ids, inference_info)
  File "/home/qtr/anaconda3/envs/petals/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/qtr/anaconda3/envs/petals/lib/python3.10/site-packages/petals/server/backend.py", line 92, in inference_step
    self._reorder_cache_inplace(cache_tensors, hypo_ids)
  File "/home/qtr/anaconda3/envs/petals/lib/python3.10/site-packages/petals/server/backend.py", line 102, in _reorder_cache_inplace
    cache_tensor[...] = cache_tensor[hypo_ids]  # in-place reorder cache by hypo ids
RuntimeError: indices should be either on cpu or on the same device as the indexed tensor (cuda:1)
Jan 31 08:16:11.407 [INFO] rpc_inference.close(blocks=7:13, remote_peer=...AM4oxF)

```

**Comment by quarterturn** [comment]
I'm connecting now as:
```
Feb 01 04:23:43.825 [INFO] Connecting to the public swarm, peer_id = 12D3KooWNM381VzV6odc63MiY5QLGK9DFFBa3jjwRwyza9mAhTZg
```

with the following code changes:

```
     98     def _reorder_cache_inplace(self, cache_tensors: torch.Tensor, hypo_ids: torch.Tensor):
     99         """If hypo_ids is specified, reorder elements of each cache tensor in-place by taking indices from hypo_ids"""
    100         if not is_dummy(hypo_ids):
    101             for cache_tensor in cache_tensors:
    102                 cache_tensor[...] = cache_tensor[hypo_ids.cpu()]  # in-place reorder cache by hypo ids
```

I'm thinking on my particular GPU it's faster to run less blocks in 16-bit mode vs more in 8-bit. We'll see hopefully.

**Comment by quarterturn** [comment]
It seems stable. I'm not sure if I've processed any work yet, but no errors so far. I'll close this tomorrow if there's no further exceptions.

**Comment by quarterturn** [comment]
No further problems seen. Closing.

**Comment by borzunov** [comment]
I have created a PR with the fix: #249
